lib/websocket/websocket.js

```diff
@@ -57,38 +58,48 @@ class WebSocket extends EventTarget {
     url = webidl.converters.USVString(url)
     protocols = options.protocols
 
-    // 1. Let urlRecord be the result of applying the URL parser to url.
+    // 1. Let baseURL be this's relevant settings object's API base URL.
+    const baseURL = getGlobalOrigin()
+
+    // 1. Let urlRecord be the result of applying the URL parser to url with baseURL.
     let urlRecord
 
     try {
-      urlRecord = new URL(url)
+      urlRecord = new URL(url, baseURL)
     } catch (e) {
-      // 2. If urlRecord is failure, then throw a "SyntaxError" DOMException.
+      // 3. If urlRecord is failure, then throw a "SyntaxError" DOMException.
       throw new DOMException(e, 'SyntaxError')
     }
 
-    // 3. If urlRecord’s scheme is not "ws" or "wss", then throw a
-    //    "SyntaxError" DOMException.
+    // 4. If urlRecord’s scheme is "http", then set urlRecord’s scheme to "ws".
+    if (urlRecord.protocol === 'http:') {
+      urlRecord.protocol = 'ws:'
+    } else if (urlRecord.protocol === 'https:') {
+      // 5. Otherwise, if urlRecord’s scheme is "https", set urlRecord’s scheme to "wss".
+      urlRecord.protocol = 'wss:'
+    }
+
+    // 6. If urlRecord’s scheme is not "ws" or "wss", then throw a "SyntaxError" DOMException.
```

Comment:
The next line triggers the protocol getters, and can just be replaced buy a boolean variable. The getters of URLs splices the string, and avoiding them is a good practice imho.